x/zkauth/types/zkauth_inputs.go

```diff
@@ -0,0 +1,199 @@
+
+	"github.com/iden3/go-iden3-crypto/poseidon"
+	"github.com/pkg/errors"
+
+	sdkerrors "github.com/Finschia/finschia-sdk/types/errors"
+)
+
+const MaxHeaderLen = 992
+const MaxExtIssLen = 32
+const PackWidth = 248
+
+// circom constants from main.circom / https://zkrepl.dev/?gist=30d21c7a7285b1b14f608325f172417b
+// template RSAGroupSigVerify(n, k, levels) {
+// component main { public [ modulus ] } = RSAVerify(121, 17)
+// component main { public [ root, payload1 ] } = RSAGroupSigVerify(121, 17, 30)
+
+const CircomBigintN = 121
+const CircomBigintK = 17
+
+func SplitToTwoFrs(ephPkBytes []byte) []*big.Int {
+	mid := len(ephPkBytes) - 16
+	first := new(big.Int).SetBytes(ephPkBytes[0:mid])
+	second := new(big.Int).SetBytes(ephPkBytes[mid:])
+
+	return []*big.Int{first, second}
+}
+
+func ToField(val string) (*big.Int, bool) {
+	return new(big.Int).SetString(val, 10)
+}
+
+func reverse[T any](s []T) {
+	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
+		s[i], s[j] = s[j], s[i]
+	}
+}
+
+// ChunkArray splits an array into chunks of a specified size.
+func ChunkArray(array []byte, chunkSize int) [][]byte {
+	// The length of the array divided by the chunk size is rounded up and used as the number of chunks.
+	chunkCount := int(math.Ceil(float64(len(array)) / float64(chunkSize)))
+	chunks := make([][]byte, chunkCount)
+
+	revArray := make([]byte, len(array))
+	copy(revArray, array)
+	reverse(revArray)
+
+	// split the revert array into chunks, and revert each chunk again.
+	for i := 0; i < chunkCount; i++ {
+		start := i * chunkSize
+		end := start + chunkSize
+		if end > len(revArray) {
+			end = len(revArray)
+		}
+		chunk := make([]byte, end-start)
+		copy(chunk, revArray[start:end])
+		reverse(chunk)
+		chunks[i] = chunk
+	}
+
+	// return the final chunk array reverted.
+	reverse(chunks)
+	return chunks
+}
+
+// BytesBEToBigInt converts a slice of bytes to a big.Int.
+func BytesBEToBigInt(bytes []byte) *big.Int {
+	if len(bytes) == 0 {
+		return big.NewInt(0)
+	}
+
+	return new(big.Int).SetBytes(bytes)
+}
+
+// HashASCIIStrToField hashes an ASCII string to a field element.
+func HashASCIIStrToField(val string, maxSize int) (*big.Int, error) {
+	bytes := []byte(val)
+	if len(bytes) > maxSize {
+		return nil, errors.New("String is longer than allowed size")
+	}
+
+	// Padding with zeroes
+	strPadded := make([]byte, maxSize)
+	copy(strPadded, bytes)
+
+	const chunkSize = PackWidth / 8
+	chunks := ChunkArray(strPadded, chunkSize)
+	packed := make([]*big.Int, 0, len(chunks))
+	for _, chunk := range chunks {
+		byteChunk := make([]byte, len(chunk))
+		copy(byteChunk, chunk)
+		packed = append(packed, BytesBEToBigInt(byteChunk))
+	}
+
+	hash, err := PoseidonHash(packed)
+	if err != nil {
+		return nil, err
+	}
+
+	return hash, nil
+}
+
+// CircomBigIntToChunkedBytes converts a big integer to a slice of chunked *big.Int.
+func CircomBigIntToChunkedBytes(num *big.Int) []*big.Int {
+	bytesPerChunk, numChunks := CircomBigintN, CircomBigintK
```

Comment:
Checking the movement, it seems that this is actually bitsPerChunk.

Reply:
Sorry, `bytesPerChunk` is right.

Reply:
msk will be 2^bytesPerChunk - 1.
To follow the name bytesPerChunk, `msk := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), uint(bytesPerChunk)*8), big.NewInt(1))`.
In other words, this is not 121 bytes but 121 bits, so I think bitsPerChunk is correct, but am I misunderstanding something?
https://github.com/Finschia/finschia-sdk/pull/1239/files#diff-b762dc12df3343605c06f48c92dc91e6ccaba97a6dfe9d31db88a1ee7f0ca7acR99

Reply:
Yes, I referred the original source's type, but as you said it's wrong. I'll fix it. Thank you.